Replace Slime.Parser with parser generated from PEG (using neotoma)

Rakoth opened the pull request:
### This is experimental implementation of a parser using [PEG](https://en.wikipedia.org/wiki/Parsing_expression_grammar)

First thing to look at is grammar [src/slime_parser.peg](https://github.com/Rakoth/slime/blob/master/src/slime_parser.peg)
Then take a look at new preprocessor [lib/slime/parser/preprocessor.ex](https://github.com/Rakoth/slime/blob/master/lib/slime/parser/preprocessor.ex) which main goal is to add indents/dedents where needed (I get the ideas from [here](https://github.com/sirthias/parboiled/wiki/Indentation-Based-Grammars))
Finally take a look at transform callback module [lib/slime/parser/transform.ex](https://github.com/Rakoth/slime/blob/master/lib/slime/parser/transform.ex) This is where parsed AST gets transformed into useful one. Its current goal is to produce tokens exactly as our current parser do. But it can be easily modified to produce nested structures as `Slime.Tree` do.

Things to discuss:
* [ ] Performance issue.
  The bummer is that this parser is ~10x slower than our current parser. But I believe grammar might be optimized to reduce the performance gap. See benchmark [bench/example_parse_bench.exs](https://github.com/Rakoth/slime/blob/master/bench/example_parse_bench.exs).
* [x] Handle embedded engine code with inconsistent indents.
  There is currently a [pending test](https://github.com/Rakoth/slime/blob/master/test/rendering/embedded_engine_test.exs#L44) case about this issue. The possible solution is to make preprocessor smart enough to skip embedded engine lines and do not add indents/dedents.
* [x] Make `:attr_list_delims` work again.
  There is [penging test case](https://github.com/Rakoth/slime/blob/master/test/rendering/attributes_test.exs#L72)
  The possible solutions are:
  - Generate slime_parser.peg file from a template and add only supported delimiters into grammar
  - Pass original iolist with transformed AST through and in `transform(:wrapped_attributes, ...)` based on config options chose ether transformed attributes or the original input.
  - Drop support :(
* [x] Raise an error on unmatched attrs delimiters.
  (When there is a space separating tag head and wrapped attributes list). [Pending test](https://github.com/Rakoth/slime/blob/master/test/parser_test.exs#L52). While it is currently supported and is supported by original slim, I found it hard to manage using PEG, such attributes now parsed as inline text after wrapped attributes check is failed.
* [ ] Grammar and transform can by simplified.
  - if we remove empty lines in preprocessor, but then we need to map errors to original input lines.
  - other suggestions?

Things to do next:
* [x] Since preprocessor now adds symbols to original input, it is necessary to map errors position to original input. And provide useful error messages on unexpected indents/dedents.
* [ ] Parse nested tags structure, remove Slime.Tree
* [ ] PR to neotoma project to support Elixir module as transform_module option?
* [ ] Separate static code and dynamic elixir insertions in strings with interpolation
* [ ] Compile to function omitting eex representation (great deal is done by @lpil [here](https://github.com/lpil/jot))

Closes #106 #32 #123

@doomspork, @henrik, @lpil, @bozydar I need your thoughts on this, guys.

Rakoth committed:
Replace Slime.Parser with parser generated from PEG

doomspork commented:
Thank you for this @Rakoth, this is awesome!  Sorry for the delay getting back to you, I've been swapped with work and family things.

Could we sync up on Slack and you give me a rough run down of things?  This is a new approach for me.

doomspork commented:
@Rakoth once the merge conflicts are resolved is this safe to merge?  Will we retain the same level of functionality we currently have?

Rakoth committed:
Fix support for :attr_list_delims config option

This introduces new compile-time dependency: neotoma parser

Rakoth committed:
Support leading and  trailing spaces for code in peg parser

Rakoth committed:
Fix code style, compilation and credo warnings

Rakoth committed:
Handle inconsistent indents in embedded engine body.

Also with this commit slime preprocessor will raise an error on
inconsistent indentation as ruby slim does.

Rakoth committed:
Show parsing errors with line and column number

Rakoth committed:
Handle nested inline tags in peg parser

Rakoth committed:
Add leading and trailing spaces to inline tags

Rakoth committed:
Parse multi-line elixir code with inconsistent indentations

Rakoth committed:
Run all parser tests with new implementation

Rakoth committed:
Remove unused code
